aiorospy/src/aiorospy/helpers.py

```diff
@@ -98,10 +109,50 @@ def _task_done_callback(self, task):
                     timeout=period)
                 return result
             except asyncio.TimeoutError:
-                sink(msg)
+                if iscoroutinefunction_or_partial(do):
+                    await do(*args, **kwargs)
+                else:
+                    do(*args, **kwargs)
             except asyncio.CancelledError:
                 task.cancel()
                 await task
                 raise
     else:
         return await awaitable
+
+
+async def log_during(awaitable, msg, period, sink=logger.info):
+    """ Convenience function to repeatedly log an line, while some task has not completed. """
+    return await do_while(awaitable, period, sink, msg)
+
+
+class ChildCancelled(asyncio.CancelledError):
+    pass
+
+
+async def detect_cancel(task):
+    """ asyncio makes it very hard to distinguish an inner cancel from an outer cancel.
```

Comment:
I think my honeymoon phase with asyncio is over. But I'm not sure there's a better alternative anyway.

Reply:
![image](https://user-images.githubusercontent.com/1694370/75694218-2b441580-5c76-11ea-89e8-f4951cf8fab7.png)